### PR TITLE
netlink: return EBADF on socket calls after Linux conn closed

### LINKED ISSUE #92: Closing a Netlink socket with ongoing Receive() call
Hi Matt

I'm trying to cleanly terminate a Netlink socket listening on a multicast group.

```go
func main() {
	conn, err := netlink.Dial(0xc, nil) // NETLINK_NETFILTER
	if err != nil {
		log.Fatal(err)
	}

	err = conn.JoinGroup(10) // Non-existent multicast group (Receive() will never return)
	if err != nil {
		log.Fatal(err)
	}

	recvChan := make(chan []netlink.Message)

	// Start a blocking receive
	go func() {
		recv, err := conn.Receive()
		if err != nil {
			log.Fatal(err)
		}
		recvChan <- recv
	}()

	// Make sure goroutine enters Receive()
	// time.Sleep(50 * time.Millisecond)

	err = conn.Close()
	if err != nil {
		log.Fatal(err)
	}

	log.Println(<-recvChan)
}
```

This uncovers two issues:

1. A race condition: running this program multiple times in sequence yields different results. One is a panic:

```
panic: send on closed channel
goroutine 4 [running]:
github.com/mdlayher/netlink.(*sysSocket).do(0xc4200bc0a0, 0xc4200f6000)
    /home/timo/go/src/github.com/mdlayher/netlink/conn_linux.go:356 +0xa8
```

.. that can be avoided by closing the channel after the WaitGroup merges. This causes a deadlock in the test suite, so this might invalidate some assumptions you've made in other places.

```go
func (s *sysSocket) Close() error {
	...
	s.wg.Wait()
	close(s.funcC)
```

2. `Receive()` is non-interruptible. (uncomment the `Sleep()` in the example)

If there's no valid Netlink message coming in (like in the example above), `Receive()` will hang forever. I've tried closing the fd from the main thread by executing `unix.Close()` directly instead of queueing it with `s.do()`, but this (expectedly) doesn't interrupt the syscall either. It will only yield when a message is read from the socket.

I know this is not directly an issue with the library itself, but it would be pleasant to have a solution or workaround. Is there anything you could think of? A quick search hinted that syscalls can be interrupted by signals, not sure if this is something we could implement.

### PR DESCRIPTION
**Pull request by mdlayher**
Fixes #92.

### CODE REVIEW

**Comment by florianl** [comment]
works fine for me

File: conn_linux.go [hunk 1]
```diff
@@ -416,7 +449,14 @@ func (s *sysSocket) Getsockname() (unix.Sockaddr, error) {
 
 	return sa, err
 }
+
 func (s *sysSocket) Recvmsg(p, oob []byte, flags int) (int, int, int, unix.Sockaddr, error) {
+	s.mu.RLock()
+	defer s.mu.RUnlock()
```
[the remark this answers is not in the document]
Comment (florianl) [reply conn_linux.go]: It might have an impact. Personally, I prefer stability over performance.

[the remark this answers is not in the document]
Comment (ti-mo) [reply conn_linux.go]: @florianl I fail to see how stability is relevant here; I care about performance, hence my comment. Of course leaving a mutex locked is a problem, but I expect to be able to call `Recvmsg`  tens of thousands of times per second without it becoming the most expensive operation in the call chain. Please read https://bytes-and-bites.com/posts/defer-go-performance.

@mdlayher Unless a `defer`-less alternative can be provided for receiving large amounts of events from multicast groups, I don't think this is the way to go. This function is in the critical path before fanout can happen to decoder goroutines. Do you have any existing benchmarks that include this path to make sure no performance regressions are introduced?

**Comment by mdlayher** [comment]
I'm not interested in speculating about the performance or lack of performance in a given solution, without concrete code to test assumptions with.

I'm merging this as it provides a simple solution that works. Performance issues can be addressed as needed after concrete evidence is found. I don't have time to do any further discovery in this area.

**Pull request merged**